File: acode-mac/ACode/Sources/Services/SkillsService.swift
import Foundation

/// 技能管理服务
/// 将技能持久化到本地 JSON 文件，并根据启用状态写入对应 AI 工具的配置目录
final class SkillsService {

    static let shared = SkillsService()

    // MARK: - Storage Path

    private var skillsDir: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        return base.appendingPathComponent("ACode/skills", isDirectory: true)
    }

    private var indexPath: URL {
        skillsDir.appendingPathComponent("index.json")
    }

    // MARK: - Claude CLAUDE.md path

    private var homeDir: URL {
        FileManager.default.homeDirectoryForCurrentUser
    }

    /// ~/.claude/CLAUDE.md — Claude Code 的自定义指令文件
    private var claudeInstructionsPath: URL {
        homeDir.appendingPathComponent(".claude/CLAUDE.md")
    }

    /// ~/.codex/instructions.md — Codex CLI 的自定义指令文件
    private var codexInstructionsPath: URL {
        homeDir.appendingPathComponent(".codex/instructions.md")
    }

    /// ~/.gemini/GEMINI.md — Gemini CLI 的自定义指令文件
    private var geminiInstructionsPath: URL {
        homeDir.appendingPathComponent(".gemini/GEMINI.md")
    }

    // MARK: - CRUD

    /// 获取所有技能
    func listSkills() -> [Skill] {
        guard let data = try? Data(contentsOf: indexPath),
              let skills = try? JSONDecoder().decode([Skill].self, from: data) else {
            return []
        }
        return skills
    }

    /// 保存技能（新建或更新）
    func saveSkill(_ skill: Skill) throws {
        var skills = listSkills()

        if let idx = skills.firstIndex(where: { $0.id == skill.id }) {
            skills[idx] = skill
        } else {
            skills.append(skill)
        }

        try persistSkills(skills)
        syncInstructionFiles(skills: skills)
    }

    /// 删除技能
    func deleteSkill(_ skill: Skill) {
        var skills = listSkills()
        skills.removeAll { $0.id == skill.id }
        try? persistSkills(skills)
        syncInstructionFiles(skills: skills)
    }

    /// 切换技能在指定 app 的启用状态
    func toggleSkillApp(_ skill: Skill, app: String, enabled: Bool) {
        var skills = listSkills()
        guard let idx = skills.firstIndex(where: { $0.id == skill.id }) else { return }
        if enabled {
            skills[idx].enabledApps.insert(app)
        } else {
            skills[idx].enabledApps.remove(app)
        }
        try? persistSkills(skills)
        syncInstructionFiles(skills: skills)
    }

    // MARK: - Persistence

    private func persistSkills(_ skills: [Skill]) throws {
        try FileManager.default.createDirectory(at: skillsDir, withIntermediateDirectories: true)
        let data = try JSONEncoder().encode(skills)
        try data.write(to: indexPath)
    }

    // MARK: - Sync to AI Tool Config Files

    /// 将启用的技能合并写入各工具的指令文件
    private func syncInstructionFiles(skills: [Skill]) {
        let marker = "<!-- ACode Skills -->"
        let endMarker = "<!-- /ACode Skills -->"

        syncInstructionFile(
            path: claudeInstructionsPath,
            skills: skills.filter { $0.enabledApps.contains("claude") },
            marker: marker,
            endMarker: endMarker
        )

        syncInstructionFile(
            path: codexInstructionsPath,
            skills: skills.filter { $0.enabledApps.contains("codex") },
            marker: marker,
            endMarker: endMarker
        )

        syncInstructionFile(
            path: geminiInstructionsPath,
            skills: skills.filter { $0.enabledApps.contains("gemini") },
            marker: marker,
            endMarker: endMarker
        )
    }

    /// 在指令文件中替换 ACode Skills 区块
    private func syncInstructionFile(path: URL, skills: [Skill], marker: String, endMarker: String) {
        // 确保父目录存在
        let parent = path.deletingLastPathComponent()
        try? FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)

        var existing = (try? String(contentsOf: path, encoding: .utf8)) ?? ""

        // 移除旧的 ACode Skills 区块
        if let startRange = existing.range(of: marker),
           let endRange = existing.range(of: endMarker) {
            let fullRange = startRange.lowerBound..<endRange.upperBound
            existing.removeSubrange(fullRange)
            // 清理多余空行
            existing = existing.trimmingCharacters(in: .whitespacesAndNewlines)
        }

        // 如果没有技能需要写入，只保留清理后的内容
        if skills.isEmpty {
            if existing.isEmpty {
                // 如果文件完全为空，不写入
                return
            }
            try? existing.write(to: path, atomically: true, encoding: .utf8)
            return
        }

        // 构建新的技能区块
        var block = "\n\n\(marker)\n"
        for skill in skills {
            block += "## \(skill.name)\n"
            if !skill.description.isEmpty {
                block += "> \(skill.description)\n"
            }
            block += "\n\(skill.content)\n\n"
        }
        block += "\(endMarker)"

        existing += block

        try? existing.write(to: path, atomically: true, encoding: .utf8)
    }
}

File: acode-windows/lib/services/skills_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as p;

import '../models/skill.dart';

/// 技能管理服务 — 持久化 JSON + 写入 CLI 指令文件
class SkillsService {
  static final SkillsService shared = SkillsService._();
  SkillsService._();

  String get _skillsFilePath {
    final home = Platform.environment['USERPROFILE'] ?? Platform.environment['HOME'] ?? '';
    return p.join(home, '.acode', 'skills.json');
  }

  /// 获取所有技能
  List<Skill> listSkills() {
    try {
      final file = File(_skillsFilePath);
      if (!file.existsSync()) return [];
      final json = jsonDecode(file.readAsStringSync());
      if (json is List) {
        return json.map((e) => Skill.fromMap(e as Map<String, dynamic>)).toList();
      }
    } catch (_) {}
    return [];
  }

  /// 保存技能
  void saveSkill(Skill skill) {
    final skills = listSkills();
    final idx = skills.indexWhere((s) => s.id == skill.id);
    if (idx >= 0) {
      skills[idx] = skill;
    } else {
      skills.add(skill);
    }
    _persistSkills(skills);
    _syncToTools(skills);
  }

  /// 删除技能
  void deleteSkill(Skill skill) {
    final skills = listSkills();
    skills.removeWhere((s) => s.id == skill.id);
    _persistSkills(skills);
    _syncToTools(skills);
  }

  /// 切换技能对某个应用的启用状态
  void toggleSkillApp(Skill skill, String app, bool enabled) {
    final newApps = Set<String>.from(skill.enabledApps);
    if (enabled) {
      newApps.add(app);
    } else {
      newApps.remove(app);
    }
    saveSkill(skill.copyWith(enabledApps: newApps));
  }

  /// 持久化技能列表到 JSON 文件
  void _persistSkills(List<Skill> skills) {
    try {
      final file = File(_skillsFilePath);
      final dir = file.parent;
      if (!dir.existsSync()) {
        dir.createSync(recursive: true);
      }
      file.writeAsStringSync(
        const JsonEncoder.withIndent('  ').convert(
          skills.map((s) => s.toMap()).toList(),
        ),
      );
    } catch (_) {}
  }

  /// 同步技能到各 CLI 工具的指令文件
  void _syncToTools(List<Skill> skills) {
    _syncToClaude(skills);
    _syncToCodex(skills);
    _syncToGemini(skills);
  }

  void _syncToClaude(List<Skill> skills) {
    final home = Platform.environment['USERPROFILE'] ?? Platform.environment['HOME'] ?? '';
    if (home.isEmpty) return;

    _syncInstructionFile(
      p.join(home, '.claude', 'CLAUDE.md'),
      skills.where((s) => s.enabledApps.contains('claude')).toList(),
    );
  }

  void _syncToCodex(List<Skill> skills) {
    final home = Platform.environment['USERPROFILE'] ?? Platform.environment['HOME'] ?? '';
    if (home.isEmpty) return;

    _syncInstructionFile(
      p.join(home, '.codex', 'instructions.md'),
      skills.where((s) => s.enabledApps.contains('codex')).toList(),
    );
  }

  void _syncToGemini(List<Skill> skills) {
    final home = Platform.environment['USERPROFILE'] ?? Platform.environment['HOME'] ?? '';
    if (home.isEmpty) return;

    _syncInstructionFile(
      p.join(home, '.gemini', 'GEMINI.md'),
      skills.where((s) => s.enabledApps.contains('gemini')).toList(),
    );
  }

  /// 在指令文件中替换 ACode Skills 区块（与 Mac 版对齐）
  void _syncInstructionFile(String filePath, List<Skill> skills) {
    const marker = '<!-- ACode Skills -->';
    const endMarker = '<!-- /ACode Skills -->';

    try {
      final file = File(filePath);
      final dir = file.parent;
      if (!dir.existsSync()) dir.createSync(recursive: true);

      var existing = file.existsSync() ? file.readAsStringSync() : '';

      // 移除旧的 ACode Skills 区块（兼容旧格式 Start/End）
      existing = _removeSkillBlock(existing, marker, endMarker);
      existing = _removeSkillBlock(existing, '<!-- ACode Skills Start -->', '<!-- ACode Skills End -->');
      existing = existing.trim();

      // 如果没有技能需要写入，只保留清理后的内容
      if (skills.isEmpty) {
        if (existing.isNotEmpty) file.writeAsStringSync(existing);
        return;
      }

      // 构建新的技能区块
      final buffer = StringBuffer('\n\n$marker\n');
      for (final skill in skills) {
        buffer.writeln('## ${skill.name}');
        if (skill.description.isNotEmpty) {
          buffer.writeln('> ${skill.description}');
        }
        buffer.writeln();
        buffer.writeln(skill.content);
        buffer.writeln();
      }
      buffer.write(endMarker);

      existing += buffer.toString();
      file.writeAsStringSync(existing);
    } catch (_) {}
  }

  String _removeSkillBlock(String content, String startMarker, String endMarker) {
    final startIdx = content.indexOf(startMarker);
    final endIdx = content.indexOf(endMarker);
    if (startIdx >= 0 && endIdx >= 0) {
      return content.substring(0, startIdx) + content.substring(endIdx + endMarker.length);
    }
    return content;
  }
}
